weather_report/main.py

```diff
@@ -0,0 +1,172 @@
+"""
+This file performs all the operations mentioned in readme for weather man problem
+"""
+import csv
+import os
+import argparse
+from datetime import datetime
+
+
+# utility functions =====================
+def year_boundary_check(check_year) -> bool:
+    """
+    Return true if year is between 2004 and 2016
+    and false otherwise
+    """
+    if check_year and 2004<=check_year<=2016:
+        return True
+    return False
+
+def year_month_validation_decorator(func):
+    """
+    Validate if the year and month given are in the right format
+    """
+    def wrapper(data,report_month):
+        date=report_month.split('/')
+        print(f"\n=== {func.__name__.replace('_', ' ').title()} for {report_month} ===")
+        if len(date) == 2 and all(value != '' for value in date):
+            r_year, r_month = map(int, date)
+            if year_boundary_check(r_year) and 1 <= r_month <= 12:
+                func(data, r_year, r_month)
+            else:
+                print("This year or month data is not available")
+        else:
+            print("Please give year and month in correct format -- yyyy/mm")
+    return wrapper
+
+
+def parse_weather_data(weather_row) -> dict:
+    """
+    parse the weather columns to respective data types
+    """
+    parsed_weather_row={}
+    for key,value in weather_row.items():
+        if key in ('PKT','PKST'):
+            parsed_weather_row['PKT'] = datetime.strptime(value,"%Y-%m-%d").date()
+        elif key ==' Events':
+            parsed_weather_row[key.strip()]=value
+        else:
+            #key.strip() remove extra spacing in key
+            parsed_weather_row[key.strip()] = 0.0 if value=='' else float(value)
+    return parsed_weather_row
+
+def read_weather_data() -> list:
+    """
+    Read the data files from weatherfiles folder and return it
+    """
+    weather_data_dir=os.path.join(os.path.dirname(__file__),'weatherfiles')
+    data_files=list(os.listdir(weather_data_dir))
+    weather_data_list=[]
+    for file_path in data_files:
+        with open(os.path.join(weather_data_dir,file_path),'r',encoding='latin-1') as f:
```

Comment:
Yes, it is. Otherwise I am getting following error:

```
csv_reader = csv.DictReader(x.replace("\0", "") for x in f)
File "/Users/raiqarasool/miniforge3/lib/python3.9/codecs.py", line 322, in decode
(result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 3131: invalid start byte
```